Primary/Secondary aware connection handling

mintsoft opened the pull request:
@roji As I just mentioned, working towards: https://github.com/npgsql/npgsql/issues/732

The core functionality I'm aiming for here is:

- [x] Application should be able to give a Host=a,b,c connection string and the driver will connect to each of them in-turn and use the first server that works
- [x] Application should be able to set ServerType=Primary to guarantee that a connection to a writable server will be handed out if one is available and error otherwise
- [x] Application should be able to set ServerType=Secondary to guarantee (as close as possible) that a connection to a readonly server will be handed out (whether this is a actually a guarantee, or a preferSecondary concept, I'm not sure yet)
- [x] Application with connectionstring of `Host=PG1,PG2;ServerType=master` in the event of PG1 going down, and PG2 becoming primary should be able to continue working without requiring a hard recycle/restart.
- [x] ConnectionPool (min, max, pruning etc) should be as close to the current behaviour as possible

I've given this a test in a few simple failover scenarios and it seems to reconnect etc, however I've not done any "to destruction" testing yet.

I've ripped out the global state cache ideal that we were discussing in favour of each `NpgsqlConnector` knowing what type of server it's connected to. When `NpgsqConnection` pulls the connector out of the pool, it checks it against whether it's what it "wants" and reconnects if it isn't which is much simpler. 

I do have an unsolved problem with this however, if a user has connectionpool enabled, but not keepalives; there will be no point for each connector to have its connected state updated if the server changes from secondary to primary (for example) and I can't see any mechanism for it to ever be updated unless the connection itself gets reset at some point. I guess https://www.postgresql.org/message-id/flat/1700970.cRWpxnom9y@hammer.magicstack.net might fix that?

flyingpie commented:
@mintsoft Great work so far, it's working like a charm, for at least a large part of the scenarios :)

A problem I'm having though.

When a connection to a host fails, a call to Break() is made, to basically undo the current connection attempt:
https://github.com/mintsoft/npgsql/blob/failover-advanced/src/Npgsql/NpgsqlConnector.cs#L436
https://github.com/mintsoft/npgsql/blob/failover-advanced/src/Npgsql/NpgsqlConnector.cs#L442

This is problematic for when a host is not reachable at all (eg. machine not reachable or Postgres server crashed entirely). Within the Break() method, the "Connection"-property is set to "null", which causes problems later on, within "LoadDatabaseInfo()":
https://github.com/mintsoft/npgsql/blob/failover-advanced/src/Npgsql/NpgsqlConnector.cs#L1442

- LoadDatabaseInfo() passes a reference to the connection (which can be null at that point):
https://github.com/mintsoft/npgsql/blob/failover-advanced/src/Npgsql/NpgsqlConnector.cs#L457

- ConnectionTypeMapper passes the reference to UnknownTypeHandler:
https://github.com/mintsoft/npgsql/blob/failover-advanced/src/Npgsql/TypeMapping/ConnectorTypeMapper.cs#L62

- UnknownTypeHandler passes the reference to TextHandler:
https://github.com/mintsoft/npgsql/blob/failover-advanced/src/Npgsql/TypeHandlers/UnknownTypeHandler.cs#L22

- TextHandler tries to dereference the connection, where it throws an exception:
https://github.com/mintsoft/npgsql/blob/failover-advanced/src/Npgsql/TypeHandlers/TextHandler.cs#L71

Since the exception gets thrown within a catch-block, it kills the entire connection process.

For my local tests, I commented out the calls to Break(), though this is probably not the greatest solution.

Not entirely sure what the correct way of doing this would be, but feel free to let me know if I can help out.

mintsoft commented:
Hiya @flyingpie 

Thanks for testing that! That's really useful information, I'll address that shortly. I had a feeling there might end up being a problem with that reset mechanism, glad you found it :)

mintsoft commented:
@flyingpie Actually I've just tried that in my little test program, connecting to a non existant host like:
```
using(var con = new NpgsqlConnection("Host=172.30.0.24;username=robtest_login;password=robtest_password;Database=robtest;TargetServerType=Primary;Pooling=true;KeepAlive=5;Enlist=false;Minimum Pool Size=4;");
con.Open();
```
and I get a `TimeoutException` as I'd expect:
```
Npgsql.NpgsqlException (0x80004005): Exception while connecting ---> System.TimeoutException: Timeout during connection attempt
   at Npgsql.NpgsqlConnector.Connect(NpgsqlTimeout timeout) in C:\Users\Rob\source\repos\npgsql\src\Npgsql\NpgsqlConnector.cs:line 739
   at Npgsql.NpgsqlConnector.Connect(NpgsqlTimeout timeout) in C:\Users\Rob\source\repos\npgsql\src\Npgsql\NpgsqlConnector.cs:line 756
   at Npgsql.NpgsqlConnector.<RawOpen>d__155.MoveNext() in C:\Users\Rob\source\repos\npgsql\src\Npgsql\NpgsqlConnector.cs:line 681
```

Is there something different/significant about my connection string vs the one you were testing with?

flyingpie commented:
@mintsoft Thank you!

2 things:
- There is a difference in "the server does not exist" (which usually causes a timeout-exception) and "server is not hosting on this port (which usually causes a socket-exception. That mostly means that I'm getting an exception immediately, while your test causes an exception after a delay.

- The exception is thrown just fine, but it breaks the "let's try the next host"-mechanism. Since the Connection-property is set to "null", whenever the connection ultimately succeeds, it still fails with the above problem.

You should be able to get the same results if you use 2 hosts, both of which exist, but one of which doesn't actually host a Postgres instance.

Does that clear stuff up?

mintsoft commented:
@flyingpie Thats perfect, I've got that reproduced now. Thanks

hkheireddine commented:
Hey guys, we have implemented PAF for our project and could not found any.Net library that supports master and secondary servers. Any idea when this will be finalized?

mintsoft commented:
@hkheireddine I'm hoping to get this in pretty quickly, however it's not entirely dependant on me :) There's the bug raised by @flyingpie that needs fixing and there's a problem (possibly the same, possibly different) that's causing some of the tests to hang/break. Once that's done I think functionally it's ready for prime-time however I'm sure there will be some changes in the review etc. 

It's difficult to say when it'll be released. If you wanted to help out either with coding or testing though configurations, that would be much appreciated! The more environments it's been thrown at, the better it will be when it's ready

mintsoft commented:
@flyingpie I think I've just pushed a fix for that situation in https://github.com/npgsql/npgsql/pull/2778/commits/2a8b998f15bcac71f0f5d2bd66cd323fc264c933 essentially not calling Break in that situation should be fine as the state will get replaced in the next connect attempt anyway, but we might need to cleanup some connection that's the same as if the driver connects to the wrong type of server. I'm not 100% that covers all the possibilities but I think they'll come out of the woodwork.

Thanks for the feedback, much appreciated!

flyingpie commented:
@mintsoft That's nice! The socket is the only (unmanaged) thing that gets set up during the connection, right? Cause in that case, the SoftCleanup should be fine.

Just a thought, but maybe this is indicative of the need for a layer of abstraction between the actual connection and the fact that we're switching between multiple hosts. In the same way that there already is an abstraction between a _connection_ and a _pool of connections_. Just a thought though.

Nice work, I'll get the latest bits and try time out.

mintsoft commented:
@flyingpie 

> Just a thought, but maybe this is indicative of the need for a layer of abstraction between the actual connection and the fact that we're switching between multiple hosts. In the same way that there already is an abstraction between a connection and a pool of connections. Just a thought though.

Yeah I've had similiar thoughts myself. I've deliberately tried to keep this light-touch at the moment as I don't really want to mess with the ConnectionPool implementation as it's being heavily worked-on elsewhere. That said, once it's all together I'm pretty sure a much cleaner implementation will come out of the woodwork.

If you find any other issues, let me know and I'll patch them up!

mintsoft commented:
@roji  @YohDeadfall @austindrenski I think this is in a "done" state now. I've fixed up the tests and added some explicit tests for Secondary failover (these require another PG cluster which is in recovery to pass, which I've added as the environmental variable `NPGSQL_SECONDARY_HOST`).

I think (despite the number of commits), it's actually reasonably simple overall. Although I'm a little concerned about the change I had to make here to get the tests to pass: https://github.com/npgsql/npgsql/pull/2778/commits/4abc059d1242ba3813c792f9c5eecc9b65c5607a I think that *could* be a breaking change if some client code is hard-coupled to the events being fired?

Anyway, welcome any feedback you guys have got.

loop-evgeny commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -369,45 +373,77 @@ internal async Task Open(NpgsqlTimeout timeout, bool async, CancellationToken ca
+                    }
+                    catch (SocketException)
+                    {
+                        if (hosts.Last() == host) 
```
This would lead to unexpected behaviour if a host is listed multiple times. That might be an unlikely scenario, but possible: e.g. maybe host A is nearby and hosts B,C,D are far away, so I set my hosts to "A,B,C,D,A" in order to try A again just before giving up.

mintsoft replied:
Hmm that _is_ true, out of curiosity is that an actual scenario that you could see yourself using?

loop-evgeny replied:
We'd have to test it first, but maybe!

mintsoft commented:
@loop-evgeny should be resolved in https://github.com/npgsql/npgsql/pull/2778/commits/6a5cecd74f20158fba9d002c7e96d2dfb17cbeb0

YohDeadfall commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -349,6 +350,9 @@ bool IsConnected
         bool _isConnecting;
 
+        private string? ConnectedHost;
+        private NpgsqlServerStatus.ServerType ConnectedServerType;
```
nit: We don't use default access modifiers (even if I personally prefer them).

YohDeadfall commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -429,6 +466,52 @@ internal async Task LoadDatabaseInfo(NpgsqlTimeout timeout, bool async)
+                Flush();
+
+                var columnsMsg = ReadMessage();
+                var rowMsg = (DataRowMessage)(ReadMessage());
```
Use `Expect<T>` method to properly report about bugs.

YohDeadfall commented on src/Npgsql/NpgsqlServerStatus.cs:
```diff
@@ -0,0 +1,17 @@
+            Primary,
+            Secondary
+        }
+    }
```
No need in so deeply hidden enum. Move the enumerable to namespace and remove the class.

YohDeadfall commented on src/Npgsql/PregeneratedMessages.cs:
```diff
@@ -51,7 +52,8 @@ internal static byte[] Generate(NpgsqlWriteBuffer buf, string query)
-
         internal static readonly byte[] DiscardAll;
+        internal static readonly byte[] ServerIsSecondary;
+
```
nit: Extra empty line not needed here.

mintsoft commented:
Just resolved some merge issues so should still be ready to go

mintsoft commented:
@roji appreciate that you're busy, but any chance of some initial feedback on this?

roji commented:
@mintsoft sorry for the long silence - I really try my best to respond to PRs quickly, but at the moment am deep into the multiplexing work (#2852, which incidentally will need to be resolved with this at some point) as well as some vacation time. I'll do my best to give this a proper review in the coming days.

mintsoft commented on src/Npgsql/NpgsqlConnectionStringBuilder.cs:
```diff
@@ -1436,5 +1455,25 @@ public enum SslMode
+        /// <summary>
+        ///  Primary (writable) server is selected for the connection.
+        /// </summary>
+        Primary,
```
It's basically difficult to do preference with the connect/disconnect model I've got here as you'd need to have information about the preferred connections in the pool before attempting to get one.

mintsoft commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -2042,5 +2130,16 @@ internal enum DataRowLoadingMode
+    /// <summary>
+    /// Represents what state the connected server is in
+    /// </summary>
+    internal enum ServerType
```
Sure

mintsoft commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -2042,5 +2130,16 @@ internal enum DataRowLoadingMode
+    /// </summary>
+    internal enum ServerType
+    {
+        Unknown,
```
It was going to be used for the scenario where it wasn't known yet before I put the querying logic to get the type into the actual connector itself

mintsoft commented:
>The current open behavior seems to attempt to get an idle connector from the pool. If that connector happens to correspond to the type the user requested (primary, secondary), it's returned immediately - all good. However, if it isn't, that connector is returned back to the pool, and you then proceed to call OpenLong, which will open a new connection - and then you return that. First, there's no reason to expect this new connection to be of the type the user requested - we should be remembering, for each host in the connection string, what its last-known status was and establish connections accordingly (at the moment you track server type on each connector, rather than on hosts in general). Another problem is that there may already be idle connectors in the pool, of the right type - but they just happened not to be the first returned when you called TryAllocateFast. That's not good - as long as there are idle connectors (of the desired type), we should not be opening new physical connections.

Yeah, totally agree. We discussed it previously and I got the impression that you would prefer to not maintain a global Host state concept anywhere in the driver; but totally happy to revisit that. 

> In addition, if we happen to be at Max Pool Size, OpenLong will try getting an idle connector from the pool again (since it sees there are idle connectors), which may again be of the wrong type - that's not being checked right now in any way. Even it were checked, it's not clear what you'd do at that point: call OpenLong again? If the pool happens to be full of the wrong type of connectors, you'd just loop forever.

True, I was basically relying on the idle pruning to fix that; however no guarantee that users have it enabled!

>What I'm trying to say, is that the advanced version of this feature - primary/secondary connection management - is quite complex, and I don't think it's possible to do it without going into the pool and making it aware of the whole business. That would most probably mean maintaining two different idle lists (primary and secondary) inside the pool, and going to the right list based on what the user is asking for. As discussed in the issue comments, we'd also have to tackle issues such killing idle secondary connections if a primary one is requested but the pool is full and no idle primaries are available.
>However, we should definitely not do any large-scale work on the pool at this point, since I'm working in parallel on rewriting it for multiplexing support (see an early draft in #2852). The pool already looks very different than it does today (and is in fact considerably simpler in many respects), so I'd recommend holding off on this work until that is done. A good spec of the behavior and/or architecture could be done before jumping into implementation.

Cool, I agree. It seems like the best option is to extend the ConnectionPool to include host state information or similiar so we can basically have our cake and eat it.

Do you think it's likely that the work in #2852 is going to end up drastically different to where it is now? Basically I'd like to rebase this from there and start to flesh it out and see if I can provide any feedback against it that would make this feature easier to handle etc before it's merged if possible.

roji commented:
> Do you think it's likely that the work in #2852 is going to end up drastically different to where it is now? Basically I'd like to rebase this from there and start to flesh it out and see if I can provide any feedback against it that would make this feature easier to handle etc before it's merged if possible.

I'd definitely hold off on any pooling-related changes before #2852 is merged. On a positive note, it's the most high-priority thing on my plate and it's getting close to done (I have some more unpushed work), so it shouldn't take very long to complete.

mintsoft commented:
@roji I can see the first version of multiplexing in now (https://github.com/npgsql/npgsql/commit/4bd4115d9605ddf12d63b927def67c8812cf74c8). Are you expecting any fundamental changes to the pooling implementation or am I good to have a stab at reimplementing this?

roji commented:
@mintsoft yes, now would be a good time to work on anything touching the pool - I anticipate doing some minor fixes, but nothing far-reaching.

Some thoughts... I'd love to be able to ship this with 5.0, but it is obviously not a trivial feature. To make sure we go in the right direction, I'd advise putting together a quick spec of how you intend to tackle things; it doesn't have to be formal or anything - just something we can discuss together before too much code gets written. It's also a good idea to submit early draft versions and iterate, rather than working a lot and submitting one big PR in the end. I'd also be fine with not implementing this for multiplexing, for now - that should simplify the design process etc. (though we should be careful not to do anything that's fundamentally incompatible with it).

Does the above sound reasonable?

mintsoft commented:
> Some thoughts... I'd love to be able to ship this with 5.0, but it is obviously not a trivial feature. 
Yeah absolutely, I'm not too fussed on when it gets folded into a release.

>To make sure we go in the right direction, I'd advise putting together a quick spec of how you intend to tackle things; it doesn't have to be formal or anything - just something we can discuss together before too much code gets written. It's also a good idea to submit early draft versions and iterate, rather than working a lot and submitting one big PR in the end. I'd also be fine with not implementing this for multiplexing, for now - that should simplify the design process etc. (though we should be careful not to do anything that's fundamentally incompatible with it).
>
>Does the above sound reasonable?

That sounds fine in concept, however the biggest issue with producing a specification or similiar (from my perspective) is that I'm not remotely familiar with the actual architecture of the code; I can put some thoughts together but there's no guarantee that they'll make any sense relative to the implementation. In order for me to be able to see and describe how I want to implement it, I basically have to implement it anyway and then de-C# the code back into words and discuss. The way I tend to work, I'm happier throwing away/reimplementing 10+ versions of the code as I find it an easier way to express what I'm thinking vs an abstract conversation. That said, if it's the way you prefer to work I'll endeavour to oblige!

roji commented:
@mintsoft no problem, definitely don't want to get hung up on process here. If you prefer just coding it up and are willing to accept that some work will be lost, we can do it that way. I'd still recommend iteratively submitting early drafts rather than developing the whole thing though.

mintsoft commented:
Closing this in favour of https://github.com/npgsql/npgsql/pull/3098 as it's almost at the same point now anyway
